src/transform/src/normalize_lets.rs

```diff
@@ -265,9 +296,25 @@ mod support {
         {
             for (id, value) in ids.iter().zip(values.iter_mut()) {
                 refresh_types_helper(value, types)?;
-                let typ = value.typ();
-                let prior = types.insert(*id, typ);
-                assert!(prior.is_none());
+                let mut typ = value.typ();
+                if let Some(prior) = types.remove(id) {
+                    // TODO: Assert some relationship between `typ` and `prior`.
+                    for (new, old) in typ.column_types.iter_mut().zip(prior.column_types.iter()) {
+                        new.nullable = new.nullable && old.nullable
+                    }
```

Comment:
My sense is perhaps we should just not create `typ`, and instead just adopt `prior` (under the theory that it reflects our best information about types). The `RelationType` analysis is not yet `LetRec` aware, but it is based on the same logic as `typ()` and when it becomes `LetRec` aware we'll want to adopt it vs `typ()`.

Reply:
I hope it is as good or better, but it is a behavioral change and I'm already sort of spooked by the recent well-formedness failures (that changing more things has the potential to make the PR increasingly about de-mucking dark corners).